Bug 1907628: OpenStack Documentation for MachineSets with Multiple Networks (#4463)
openshift/installer: docs/user/openstack/customization.md

## Custom Subnets

Before running the installer with a custom `machinesSubnet`, there are a few things you should check. First, make sure that you have a network and subnet available. Note that the installer will create ports on this network, and some ports will be given a fixed IP addresses in the subnet's CIDR. Make sure that the credentials that you give the installer are authorized to do this! Also note that the installer will not create a router, so make sure to do this if you need a router to attach floating IP addresses to the network you want to install the nodes onto. For now, this feature does not fully support provider networks.

> Once that is ironed out, you can pass the ID of the subnet to the install config with the `machinesSubnet` option. This subnet will be set as the primarySubnet in the `machineSpec` for all nodes in this cluster. It is also where the Kubernetes endpoints of the nodes in your cluster will be published, and where the VIP ports will be created. This does add some complexity to the install process that needs to be accounted for.

Reviewer: "the install config" -> in the install-config.yaml file

> Once that is ironed out, you can pass the ID of the subnet to the install config with the `machinesSubnet` option. This subnet will be set as the primarySubnet in the `machineSpec` for all nodes in this cluster. It is also where the Kubernetes endpoints of the nodes in your cluster will be published, and where the VIP ports will be created. This does add some complexity to the install process that needs to be accounted for.

Reviewer: Do you need "This does add some complexity to the install process that needs to be accounted for."
if so, you might want to explain what the "complexity" is 
Or is the complexity the lines below?

1. The `machinesSubnet` must have DHCP enabled.
2. The CIDR of `machinesSubnet` must match the cidr of `networks.machineNetwork`.

> 3. The API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To avoid other services taking the ports assigned to the api and ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the install config to addresses that are outside of the DHCP allocation pool.

Reviewer: again "the install config" could be confusing to a new user. can you use the  "install-config.yaml file"

## Custom Subnets

> In the `install-config.yaml`, the `machinesSubnet` is the subnet where the Kubernetes endpoints of the nodes in your cluster will be published, and where the Ingress and API ports will be created. The installer creates a network and subnet for this purpose by default, however, you have the option to prevent the installer from creating those resources and overriding them with your own by setting the value of the `machinesSubnet` to the ID of an existing OpenStack subnet. To use this feature there are a few requirements:

Reviewer 2: This is an improvement! I'd probably do something like this suggestion, but do what you like. 

```suggestion
In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published.  The Ingress and API ports are created on this subnet, too.

By default, the installer creates a subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the ID of an existing OpenStack subnet. 

To use your own subnet, your cluster must meet the following requirements:
```

Thoughts:

- Do we need to indicate "network and subnet," here, or is just "subnet" sufficient?
- Use "ID" or "UUID?"

Author: It creates both a network and subnet, so its good to be clear about that. UUID might be better

> * The `machinesSubnet` must have DHCP enabled

Reviewer 2: Could punctuate these as sentences. 

Structure could also be grammatically simpler. For example:

"The subnet that is used by `machinesSubnet` has DHCP enabled."

> * The CIDR of `machinesSubnet` must match the cidr of `networks.machineNetwork`

Reviewer 2: ```suggestion
* The CIDR of `machinesSubnet` matches the CIDR of `networks.machineNetwork`
```

> * The installer user must have permission to create ports on this network, including ports with fixed IP addresses

Reviewer 2: ```suggestion
* The installer user has permission to create ports on this network, including ports that have fixed IP addresses
```

> You should also be aware of the following details:

Reviewer 2: I'd like a more specific word than "details" if possible. Limitations? Requirements?

Author: how does expectations sound?

Author: Limitations works

* If you plan to install with floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`

> * The API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To avoid other services taking the ports assigned to the api and ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.

Reviewer 2: ```suggestion
* By default, the API and Ingress VIPs try to take the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```

> * Setting `externalDNS` while setting `machinesSubnet` is invalid usage, since the installer will not modify your subnet. If you want to add a DNS to your cluster while using a custom subnet, add it to the subnet in openstack [like this](https://docs.openstack.org/neutron/rocky/admin/config-dns-res.html).

Reviewer 2: ```suggestion
* You cannot use the `externalDNS` property at the same time as a custom `machinesSubnet`. If you want to add a DNS to your cluster while using a custom subnet, [add it to the subnet in OpenStack](https://docs.openstack.org/neutron/rocky/admin/config-dns-res.html).
```

## Custom Subnets

> In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published, and the Ingress and API ports are created. By default, the installer creates a network and subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the UUID of an existing OpenStack subnet. To use this feature there are a few requirements:

Reviewer 2: Optional: 

```suggestion
In the `install-config.yaml` file, the value of the `machinesSubnet` property is the subnet where the Kubernetes endpoints of the nodes in your cluster are published. The Ingress and API ports are created on this subnet, too. By default, the installer creates a network and subnet for these endpoints and ports. Alternatively, you can use a subnet of your own by setting the value of the `machinesSubnet` property to the UUID of an existing OpenStack subnet. To use this feature, there are a few requirements:
```

Nitpicking:

At the end, "There are" is a little mushy to me. Something like "The target environment must meet the following requirements" seems closer.

Author: done

* The subnet that is used by `machinesSubnet` has DHCP enabled.
* The CIDR of `machinesSubnet` matches the CIDR of `networks.machineNetwork`.
* The installer user must have permission to create ports on this network, including ports with fixed IP addresses.

You should also be aware of the following limitations:

> * If you plan to install with floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`.

Reviewer 2: ```suggestion
* If you plan to install a cluster that uses floating IPs, the `machinesSubnet` must be attached to a router that is connected to the `externalNetwork`.
```

* The installer will not create a private network or subnet for your OpenShift machines if the `machinesSubnet` is set in the `install-config.yaml`.

> * By default, the API and Ingress VIPs by default will try to take the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, it is recommended that users set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.

Reviewer 2: ```suggestion
* By default, the API and Ingress VIPs use the .5 and .7 of your network CIDR. To prevent other services from taking the ports that are assigned to the API and Ingress VIPs, set the `apiVIP` and `ingressVIP` options in the `install-config.yaml` to addresses that are outside of the DHCP allocation pool.
```

## Custom Subnets

> Before running the installer with a custom `machinesSubnet`, there are a few things you should check. First, make sure that you have a network and subnet available. Note that the installer will create ports on this network, and some ports will be given a fixed IP addresses in the subnet's CIDR. Make sure that the credentials that you give the installer are authorized to do this! Also note that the installer will not create a router, so make sure to do this if you need a router to attach floating IP addresses to the network you want to install the nodes onto. For now, this feature does not fully support provider networks.

Reviewer 2: This and the following line would probably be more digestible as lists.
